Fixes to test framework and GitHub workflow

NightTsarina opened the pull request:
After many tries, I think I finally manage to find the right combination of
tweaks to make pytest, tox, and GitHub actions to work.

This also adds caching of the librocksdb compilation, which is a very slow
step, and splits the building of wheels and other artifacts to run them only when
pushing to main.

CC: @dato who has more experience with this than I.
CC2: Debugging this is HELL, it is turtles all the way down!

NightTsarina committed:
Fixes to test framework and GitHub workflow

After many tries, I think I finally manage to find the right combination of
tweaks to make pytest, tox, and GitHub actions to work.

This also adds caching of the librocksdb compilation, which is a very slow
step, and restricts building of wheels and other artifacts to run only when
pushing to main.

dato commented on .github/workflows/build.yml:
```diff
@@ -4,128 +4,96 @@ name: 'Build'
+        id: cache-librocksdb
         with:
-          python-version: ${{ matrix.py_ver }}
+          key: ${{ matrix.os }}-librocksdb-${{ matrix.rocksdb_ver }}
```
You would want `librocksdb-${{ matrix.os }}-${{ matrix.rocksdb_ver }` if the partial `restore-keys` are to have any effect. But...

dato commented on .github/workflows/build.yml:
```diff
@@ -4,128 +4,96 @@ name: 'Build'
+        if: steps.cache-librocksdb.outputs.cache-hit != 'true'
         run: |
           pushd /opt &&
           git clone https://github.com/facebook/rocksdb &&
           cd rocksdb &&
           git reset --hard ${{ matrix.rocksdb_ver }} &&
           CXXFLAGS='-flto -Os -s' PORTABLE=1 make shared_lib -j 4 &&
```
I haven't been able to test yet (I tried #3 for that) but, if there's a partial key match (say, for a different RocksDB version in the same operating system)… I _fear_ the git clone might fail, because the directory eists and is not empty?

Plus, what is the benefit of a partial-key match here? We could reuse the clone, but it'd also have the artifacts of a previous build, and no (visible) clean is being done. So after the reset, make would have to get right to redo all targets, correctly. (Of course that its job, but some dependency might have been missed.)

I _think_ I'd remove partial-key matches completely and, perhaps, use the standard checkout@v2 action to checkout the repository (not sure if that takes a destination_path: /opt). Or, at least, perhaps use `--depth 0 --branch ${{ rocksdb_ver }}`?

NightTsarina replied:
You are right, I completely missed this. I checked, and the checkout action does not let you specify a path outside of the workdir, so I kept the manual clone, but changed it to avoid reusing a cache from a different version.

dato reviewed (commented):
(I'm not sure where my intro went. This started with something like: _This overall LGTM, I just have some observations regarding `restore-keys`._)

NightTsarina committed:
Rework caching and building librocksdb.

Following suggestions from code review, this commit avoid re-using the working
directory from different versions of the library.

NightTsarina commented:
Comments addressed, enabled building and testing for multiple rocksdb and python versions, and all still works! PTAL?

dato reviewed (approved):
LGTM!

dato commented on .github/workflows/build.yml:
```diff
@@ -1,131 +1,99 @@
 name: 'Build'
 
-on: ['push', 'pull_request']
+on:
+  push:
```
@NightTsarina I missed the implications of this change. I think we might want to include `pull_request` here (I'll submit a PR).

Rationale: we have in the past set it to only push in repos in which we only create PRs out of branches in the same repo (not forks). This way CI for push was shared for PR, and there were no duplicate checks. But here, with forks as in #4, the checks don't run in this repo / don't show on the PR itself.

Unless you prefer some other configuration, I'll prepare a configuration with these semantics:

```
on:
  - pull_request
  - push
    - branches: main
```

NightTsarina replied:
Ah, I haven't thought of that, thanks!
